### PR TITLE
Create method of authorized transfer of all stakes between addresses

### PR DESCRIPTION
**Pull request by crazybuster**
If you made a contract change, make sure to complete the checklist below before merging it in master.

Refer to our [documentation](https://github.com/OriginProtocol/security) for more details about contract security best practices.

Contract change checklist:
  - [ ] Code reviewed by 2 reviewers. See [template](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review.md) and [example](https://github.com/OriginProtocol/security/blob/master/templates/Contract-Code-Review-Example.md).
  - [ ] Unit tests pass
  - [ ] Slither tests pass with no warning
  - [ ] Echidna tests pass if PR includes changes to OUSD contract (not automated, run manually on local)

### CODE REVIEW

**Review by franckc — COMMENTED** [review]
Nice! Took a quick pass and makes sense to me.

Could you add a migration file that we can user deploy this change?

Also bonus point for adding a small script that given the transfer agent pk as input would output the signature to communicate to the user for recovering their funds.

**Review by DanielVF — COMMENTED** [review]
1. So the use case for this is the case in which a user has had their ETH account seed phrase phished, and needs to recover their OGN without allowing the attacker who also has the account keys to transfer out the OGN as soon as it is unstaked.

This code requires the victim user to send the transaction, which requires ETH to be in the victim account, and this ETH amount will need to be several multiples larger than the amount of ETH that it would cost to transfer the ETH out. This means that if the account is being monitored by the attacker, when the user transfers ETH into the account to pay for the gas for recovering the staked OGN, then the attacker can immediately profitably steal that ETH, which blocks the recovery process.

It would seem better to use a signature from the victim, rather than a transaction. This would mean that an attacker cannot block it by stealing ETH, and sees no activity from the compromised account.

2. With the current implementation, the signature is good forever, which means that once a transfer has been authorized, it can be reused repeatedly. While in the normal use case that should not be a problem, it's still pretty sketchy. We should probably have an expiration timestamp passed along and in the signature.

3. Glad to see transferAgent required to be non-zero. That would be chaos otherwise.

**Comment by crazybuster** [comment]
> Nice! Took a quick pass and makes sense to me.
> 
> Could you add a migration file that we can user deploy this change?
> 
> Also bonus point for adding a small script that given the transfer agent pk as input would output the signature to communicate to the user for recovering their funds.



> 1. So the use case for this is the case in which a user has had their ETH account seed phrase phished, and needs to recover their OGN without allowing the attacker who also has the account keys to transfer out the OGN as soon as it is unstaked.
> 
> This code requires the victim user to send the transaction, which requires ETH to be in the victim account, and this ETH amount will need to be several multiples larger than the amount of ETH that it would cost to transfer the ETH out. This means that if the account is being monitored by the attacker, when the user transfers ETH into the account to pay for the gas for recovering the staked OGN, then the attacker can immediately profitably steal that ETH, which blocks the recovery process.
> 
> It would seem better to use a signature from the victim, rather than a transaction. This would mean that an attacker cannot block it by stealing ETH, and sees no activity from the compromised account.
> 
> 1. With the current implementation, the signature is good forever, which means that once a transfer has been authorized, it can be reused repeatedly. While in the normal use case that should not be a problem, it's still pretty sketchy. We should probably have an expiration timestamp passed along and in the signature.
> 2. Glad to see transferAgent required to be non-zero. That would be chaos otherwise.

That is a good call, @joshfraser what do you think?

**Comment by joshfraser** [comment]
Good call @DanielVF. It sounds like signature-based is the way to go.

**Comment by crazybuster** [comment]
@DanielVF @franckc @joshfraser Alright added a page /signTransfer to the dapp that will allow an account to sign the transfer. Then the use must send this signature to us and we run the /scripts/staking/transferStakes.js script with the signature to do the actual transfer.

File: dapp/src/components/EmailForm.js [hunk 1]
```diff
@@ -75,7 +75,7 @@ const EmailForm = ({ footer }) => {
           }}
           required
           value={email}
-          placeholder="Your email"
+          placeholder="Destination Address"
```
Comment (sparrowDom) [inline dapp/src/components/EmailForm.js:78]: This was probably not intended

File: dapp/src/components/SignTransferAuth.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  })
+
+  return (
+    <>
+      <div>
+        <div className="content-holder flex-grow d-flex flex-column shadow-div">
+          {active && (
+            <div>
+              {' '}
+              on {ognStaking.address} Transfer stakes from {account} to:
```
Comment (sparrowDom) [inline dapp/src/components/SignTransferAuth.js:30]: Would be cool to use `fbt` so this text is translateable

**Review by sparrowDom — COMMENTED** [review]
Nice, if Aure wants to make some mocks, I can help with translating those to frontend code if you want.

File: dapp/src/components/SignTransferAuth.js (hunk 2 above)
Comment (DanielVF) [reply dapp/src/components/SignTransferAuth.js:30]: This is probably going to be only used once or twice a quarter. Fine not to translate it, or do more work than making sure it works.

Comment (joshfraser) [reply dapp/src/components/SignTransferAuth.js:30]: Yeah, this doesn't need translations or to look pretty at all. Let's save Aure's time for more important stuff.

Comment (sparrowDom) [reply dapp/src/components/SignTransferAuth.js:30]: Makes sense, got it

File: dapp/src/components/EmailForm.js (hunk 1 above)
Comment (crazybuster) [reply dapp/src/components/EmailForm.js:78]: Yeah my bad, too many tabs Open

File: contracts/contracts/staking/SingleAssetStaking.sol [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { Initializable } from "@openzeppelin/upgrades/contracts/Initializable.so
 import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
 import { Governable } from "../governance/Governable.sol";
 import { StableMath } from "../utils/StableMath.sol";
+import "hardhat/console.sol";
```
Comment (franckc) [inline contracts/contracts/staking/SingleAssetStaking.sol:10]: remove

File: contracts/scripts/staking/signTransferAuth.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Script for verifying against etherscan.
```
Comment (franckc) [inline contracts/scripts/staking/signTransferAuth.js:1]: nit: update

File: contracts/scripts/staking/transferStakes.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+//      node signStakingPayout.js <PrivateKey> <fromAddress> <toAddress> <r> <s> <v>
+//
+//
+//
+const { ethers } = require("hardhat");
+const { utils, BigNumber, Wallet } = require("ethers");
+const { getTxOpts } = require("../../utils/tx");
+const fs = require("fs");
+
+const TEST_AGENT_PK =
```
Comment (franckc) [inline contracts/scripts/staking/transferStakes.js:14]: delete

File: contracts/scripts/staking/transferStakes.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  const wallet = new Wallet(pk, ethers.provider);
+  const proxyAddress = (await ethers.getContract("OGNStakingProxy")).address;
+  const contract = await ethers.getContractAt(
+    "SingleAssetStaking",
+    proxyAddress
+  );
+
+  const ops = await getTxOpts();
+  ops.gasLimit = 6000000;
+  ops.gasPrice = 9000000000;
```
Comment (franckc) [inline contracts/scripts/staking/transferStakes.js:27]: With the introduction of EIP 1559, I think ehter.js by default sends type 2 transaction and setting the gas price is ignored.
I would just remove all this and only set gasLimit in the call to transferStakes below.
like 
```
.transferStakes(fromAddress, toAddress, r, s, v, { gasLimit: 9000000000);
```

File: contracts/scripts/staking/signTransferAuth.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+//  - Run:
+//      node signStakingPayout.js <PrivateKey> <toAddress>
+//
+//
+//
+const { ethers } = require("hardhat");
+const { utils, BigNumber, Wallet } = require("ethers");
+const fs = require("fs");
+
+const TEST_AGENT_PK =
```
Comment (franckc) [inline contracts/scripts/staking/signTransferAuth.js:13]: remove

File: contracts/scripts/staking/signTransferAuth.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+//
+// Usage:
+//  - Run:
+//      node signStakingPayout.js <PrivateKey> <toAddress>
+//
+//
+//
+const { ethers } = require("hardhat");
+const { utils, BigNumber, Wallet } = require("ethers");
+const fs = require("fs");
```
Comment (franckc) [inline contracts/scripts/staking/signTransferAuth.js:11]: unnecessary import? pls run linter

File: contracts/scripts/staking/transferStakes.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+// Usage:
+//  - Run:
+//      node signStakingPayout.js <PrivateKey> <fromAddress> <toAddress> <r> <s> <v>
+//
+//
+//
+const { ethers } = require("hardhat");
+const { utils, BigNumber, Wallet } = require("ethers");
+const { getTxOpts } = require("../../utils/tx");
+const fs = require("fs");
```
Comment (franckc) [inline contracts/scripts/staking/transferStakes.js:12]: unnecessary import? pls run linter

**Review by franckc — APPROVED** [review]
Added  few more small comments but overall code looks good.

Let's clean this up asap and deploy!

**Comment by franckc** [comment]
A few remaining comments. Especially the one regarding the migration (not sure if the old migration format still works?)
@crazybuster do you mind taking a look? let's get this out! :)

File: contracts/scripts/staking/signTransferAuth.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
+// Script for verifying against etherscan.
+//
+// Usage:
+//  - Run:
+//      node signStakingPayout.js <PrivateKey> <toAddress>
```
Comment (sparrowDom) [inline contracts/scripts/staking/signTransferAuth.js:5]: this probably needs updating?

File: contracts/scripts/staking/transferStakes.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
+// Script for verifying against etherscan.
+//
+// Usage:
+//  - Run:
+//      node signStakingPayout.js <PrivateKey> <fromAddress> <toAddress> <r> <s> <v>
```
Comment (sparrowDom) [inline contracts/scripts/staking/transferStakes.js:5]: does this need updating as well?

**Pull request merged**